pumpkin-lib/src/engine/conflict_analysis/semantic_minimiser.rs

```diff
@@ -0,0 +1,998 @@
+use std::cmp;
+
+use super::ConflictAnalysisContext;
+use super::ConflictAnalysisResult;
+use crate::basic_types::moving_averages::MovingAverage;
+use crate::basic_types::HashSet;
+use crate::basic_types::KeyedVec;
+use crate::engine::AssignmentsInteger;
+use crate::engine::AssignmentsPropositional;
+use crate::engine::VariableLiteralMappings;
+use crate::predicate;
+use crate::predicates::IntegerPredicate;
+use crate::propagators::SparseSet;
+use crate::pumpkin_assert_simple;
+use crate::variables::DomainId;
+use crate::variables::Literal;
+
+/// Performs semantic minimisation; this minmiser attempts to remove redundant literals from
+/// nogoods.
+///
+/// For examle, if we have the nogood `[x >= 3] /\ [x >= 5] /\ ...` then we can remove `[x >= 3]`
```

Comment:
Discuss in terms in clauses, as that is more consistent with the rest of the code